lexicon/providers/hetzner.py

```diff
@@ -0,0 +1,607 @@
+            params = request.get('params', {})
+            for param in params:
+                params[param] = params[param].replace('<index>', str(page))
+            response = self._get(url, query_params=params)
+            domain_tags = Provider._filter_dom(response.text, api['filter'], True)
+            for domain_tag in domain_tags:
+                exp_domain_tag = dict(domain_tag.attrs)[api['id']['attr']]
+                domain_id = Provider._extract_domain_id(exp_domain_tag, api['id']['regex'])
+                domain = (Provider._filter_dom(domain_tag, api['domain'])
+                          .renderContents().decode('UTF-8'))
+                domains[domain] = domain_id
+                if domain == qdomain:
+                    qdomain_id = domain_id
+                    break
+            page += 1
+        if qdomain_id is None:
+            LOGGER.error('Hetzner => ID for domain %s does not exists', qdomain)
+            raise AssertionError
+        LOGGER.info('Hetzner => Get ID %s for domain %s', qdomain_id, qdomain)
+        return qdomain_id
+
+    def _get_zone(self, domain, domain_id):
+        api = self.api[self.account]
+        for request in api['zone']['GET']:
+            request = request.copy()
+            url = request.get('url', '/').replace('<id>', domain_id)
+            params = request.get('params', {})
+            for param in params:
+                params[param] = params[param].replace('<id>', domain_id)
+            response = self._get(url, query_params=params)
+        dom = Provider._filter_dom(response.text, api['filter'])
+        zone_file_filter = [{'name': 'textarea', 'attrs': {'name': api['zone']['file']}}]
+        zone_file = Provider._filter_dom(dom, zone_file_filter).renderContents().decode('UTF-8')
+        hidden = Provider._extract_hidden_data(dom)
+        zone = {'data': dns.zone.from_text(zone_file, origin=domain, relativize=False),
+                'hidden': hidden}
+        LOGGER.info('Hetzner => Get data for domain ID %s', domain_id)
+        return zone
+
+    def _post_zone(self):
+        api = self.api[self.account]['zone']
+        data = self.zone['hidden']
+        data[api['file']] = self.zone['data'].to_text(relativize=True)
+        response = self._post(api['POST']['url'], data=data)
+        if Provider._filter_dom(response.text, api['filter']):
+            LOGGER.error('Hetzner => Unable to update data for domain ID %s: Syntax error\n\n%s',
+                         self.domain_id,
+                         self.zone['data'].to_text(relativize=True).decode('UTF-8'))
+            return False
+
+        LOGGER.info('Hetzner => Update data for domain ID %s\n\n%s',
+                    self.domain_id,
+                    self.zone['data'].to_text(relativize=True).decode('UTF-8'))
+        if self.account == 'robot':
+            LOGGER.info('Hetzner => Wait 30s until Hetzner Robot has took over data...')
+            time.sleep(30)
```

Comment:
How to fasten up offline test? Now on every create|update|delete call who changes the zone file on Hetzner Robot, there will be a delay of 30s, because in live mode Hetzner need this time to took over the updated zone file.

Reply:
See my response above.

Reply:
It implies to poll until the changes are propagated, if a part of the API allow you to check this propagation.

Nevertheless you should not state a hard coded 30 seconds without any way to modify it. For any reason, the provider could take more than 30 seconds to propagate then changes. So you should put in the provider options a parameter to modify this value.

And then, you have another solution to fasten offline tests: record the modifications normally with the 30 s time sleep, then push you code along with the cassettes by overriding the parameter to a low value, like 1s. During the CI, cassettes will be replayed correctly, but without waiting 30 s each time.